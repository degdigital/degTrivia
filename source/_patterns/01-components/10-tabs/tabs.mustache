<div class="tab-group">
    <button class="button tab-button tab-button--active">Tab 1</button>
    <button class="button tab-button">Tab 2</button>
</div>