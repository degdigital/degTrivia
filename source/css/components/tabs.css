.tab-button--active {
    background-color: purple;
}