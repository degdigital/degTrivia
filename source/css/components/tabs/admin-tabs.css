.tabs {
    display: flex;
    list-style: none;
    margin-bottom: 0;
    padding-left: 0;
}

.tab {
    margin-bottom: 0;
}

.tab__button {
    font-size: 1em;
	background-color: transparent;
    border-bottom-color: transparent;
    border-left-color: color(var(--color-gray-light-40) a(46%));
    border-right-color: color(var(--color-gray-light-40) a(46%));
	border-top-color: color(var(--color-gray-light-40) a(46%));
	border-style: solid;
    border-width: .3125em;
    color: var(--color-white);
    cursor: pointer;
	font-family: var(--font-family-primary);
    margin: 0;
    padding: .5em 1.25em;
    transition: border-color var(--anim-duration-fast) var(--anim-easing);
	width: 100%;

	&.is-active {
        border-left-color: var(--color-white);
        border-right-color: var(--color-white);
        border-top-color: var(--color-white);
		outline: 0;
    }
    
    &:hover, &:focus {
        border-left-color: var(--color-gray-light-40);
        border-right-color: var(--color-gray-light-40);
        border-top-color: var(--color-gray-light-40);
    }
}

.tab-content {
    border: .3125em solid var(--color-white);
    padding: 2rem 1rem;
}