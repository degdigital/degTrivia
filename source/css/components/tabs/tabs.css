.tab-group {
    margin-bottom: 1.5em;
    text-align: center;
}

.tab__button {
    background-color: var(--color-gray-light-40);
	text-transform: none;

	&:hover, &:focus {
		background-color: var(--color-gray-light-50);
	}

	&:active {
		background-color: var(--color-white);
    }
    
    &:first-child {
        border-bottom-right-radius: 0px;
        border-top-right-radius: 0px;
    }

    &:last-child {
        border-bottom-left-radius: 0px;
        border-top-left-radius: 0px;
    }
}

.tab__button--active {
    background-color: var(--color-green);

    &:hover, &:focus {
		background-color: var(--color-green-light-20);
	}

	&:active {
		background-color: var(--color-green-light-40);
	}
}