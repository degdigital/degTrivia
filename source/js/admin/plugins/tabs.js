const tabs = function() {
	
	const triggerClass = 'tab-trigger';
	const sectionClass = 'tab-section';
	const activeClass = 'is-active';
	let triggerEls;
	let sectionEls;

	function init() {
		triggerEls = Array.from(document.querySelectorAll(`.${triggerClass}`));
		sectionEls = Array.from(document.querySelectorAll(`.${sectionClass}`));
		setInitialTab();
		bindEvents();
	}

	function bindEvents() {
		document.addEventListener('click', onDocClick);
	}

	function onDocClick(e) {
		const el = e.target;
		if (el.matches(`.${triggerClass}`)) {
			const sectionToShow = el.dataset.target;
			toggleTrigger(sectionToShow);
			toggleSection(sectionToShow);
		}
	}

	function setInitialTab() {
		const sectionFromUrl = getSectionFromUrl();
		const initialSection = sectionFromUrl || sectionEls[0].dataset.section;
		toggleTrigger(initialSection);
		toggleSection(initialSection, false);
	}

	function toggleTrigger(sectionToShow) {
		triggerEls.forEach(el => {
			const clsMethod = el.dataset.target === sectionToShow ? 'add' : 'remove';
			el.classList[clsMethod](activeClass);
		});
	}

	function toggleSection(sectionToShow, updateUrl = true) {
		sectionEls.forEach(el => {
			const clsMethod = el.dataset.section === sectionToShow ? 'add' : 'remove';
			el.classList[clsMethod](activeClass);
			if (updateUrl) {
				location.hash = sectionToShow;
			}
		});
	}

	function getSectionFromUrl() {
		const hashVal = location.hash;
		return hashVal.length > 0 ? hashVal.replace('#', '') : null;
	}

	init();

};

export default tabs;