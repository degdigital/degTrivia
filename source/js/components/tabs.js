let activeIndex;
let sectionEls;
let navItemEls;

function activateTab(newIndex) {
    //TODO: polyfill classList toggle
    sectionEls.forEach((sectionEl, index) => {
        sectionEl.classList.toggle('is-hidden', index !== newIndex);
    });
    navItemEls.forEach((navItemEl, index) => {
        navItemEl.classList.toggle('tab-button--active', index === newIndex);
    });
}

function bindEvents(containerEl) {
    containerEl.addEventListener('click', onTabClick);
}

function onTabClick(event) {
    // TODO: polyfill closest
    const navEl = event.target.closest('.tab-button');
    if (navEl) {
        const index = parseInt(navEl.getAttribute('data-tab-index'));
        activeIndex = index;
        activateTab(index);
    }
}

function destroy(containerEl) {
    if (containerEl) {
        containerEl.removeEventListener('click', onTabClick);
    }
}

export default function(containerEl) {
    if (containerEl) {
        sectionEls = [...containerEl.querySelectorAll('.tab-section')];
        navItemEls = [...containerEl.querySelectorAll('.tab-button')];
        bindEvents(containerEl);
    }

    return {
        destroy: () => destroy(containerEl)
    }
}