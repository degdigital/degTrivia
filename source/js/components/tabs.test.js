import tabs from './tabs.js';
import polyfillsForTests from '../utils/testUtils.js';

describe('tabs', () => {
    let containerEl;

    polyfillsForTests();

    beforeEach(() => {
        containerEl = document.createElement('div');
        containerEl.classList.add('tabs-container');
        containerEl.innerHTML = `
        <div class="tabs-container">
            <div class="tab-section tab-section--game" data-tab-name="Game"></div>
            <div class="tab-section tab-section--event is-hidden" data-tab-name="Event"></div>
        </div>
        `;
        document.body.appendChild(containerEl);
        tabs(containerEl);
    });

    it('should render tabs', () => {
        expect(containerEl).toMatchSnapshot();
    })

    it('should change content on tab click', () => {
        const dayTabBtn = document.querySelector('[data-tab-index="1"]');
        dayTabBtn.click();
        expect(containerEl).toMatchSnapshot();
    })

});