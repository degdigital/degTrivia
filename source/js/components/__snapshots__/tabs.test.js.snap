// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tabs should change content on tab click 1`] = `
<div
  class="tabs-container"
>
  
            
  <div
    class="tab-group"
  >
    
                
    <button
      class="button button--short tab-button"
      data-tab-index="0"
    >
      Game
    </button>
    
                
    <button
      class="button button--short tab-button tab-button--active"
      data-tab-index="1"
    >
      Day
    </button>
    
                
    <button
      class="button button--short tab-button"
      data-tab-index="2"
    >
      Event
    </button>
    
            
  </div>
  
            
  <div
    class="tab-section tab-section--game is-hidden"
  >
    Game
  </div>
  
            
  <div
    class="tab-section tab-section--day"
  >
    Day
  </div>
  
            
  <div
    class="tab-section tab-section--event is-hidden"
  >
    Event
  </div>
  
        
</div>
`;
